### PR TITLE
feat: Create interface for eval events. 

### PR DESCRIPTION
**Pull request by connyay**
## This PR

Before this change the eval services used a unexported
struct which prevented creating eval services outside of
this package.

This change creates a new IEvents interface that allows
providing custom impls of flag eval services.

### CODE REVIEW

File: flagd/pkg/service/flag-evaluation/connect_service.go
Comment (connyay) [inline flagd/pkg/service/flag-evaluation/connect_service.go:83]: Moving this to NewEventingConfiguration introduces some nasty churn on this file moving to
`iservice "github.com/open-feature/flagd/core/pkg/service"`. Could make the change if desired.

**Comment by connyay** [comment]
would like to mount the eval service in an existing server, but dont see a way I can get this working without EventingConfiguration being exported.

It is possible I am breaking the warranty seal here, but this is psuedo code of what I'm trying to do:


```go
func (f *flagdWrapper) Init(ctx context.Context) error {
	log := logging.GetLogger().With("component", "flagd.evaluation.v1").Desugar()
	reqIDLogging := false
	logger := logger.NewLogger(log, reqIDLogging)
	marshalOpts := flagdService.WithJSON(
		protojson.MarshalOptions{EmitUnpopulated: true},
		protojson.UnmarshalOptions{DiscardUnknown: true},
	)
	var (
		eval            evaluator.IEvaluator
		metricsRecorder *telemetry.MetricsRecorder
	)
	evalSvc := flagdService.NewFlagEvaluationService(
		logger, eval, nil /*eventingCfg*/, metricsRecorder,
	)
	_, flagdHandler := evaluationV1.NewServiceHandler(evalSvc, marshalOpts)

	prefix := "/flagd.evaluation.v1.Service/"
	service.MustGetHTTPServeMux(ctx).
		Handle(prefix, flagdHandler)

	return f.Initializer.Init(ctx)
}
```

File: flagd/pkg/service/flag-evaluation/eventing.go
Comment (Kavindu-Dodan) [inline flagd/pkg/service/flag-evaluation/eventing.go:10]: Can you introduce a simple interface for this so that it improves the testablity of the implementation?

File: flagd/pkg/service/flag-evaluation/connect_service.go
Comment (Kavindu-Dodan) [inline flagd/pkg/service/flag-evaluation/connect_service.go:80]: You may parameterize this to `NewConnectService()` so that the connect service works with an implementation of an interface.

**Comment by Kavindu-Dodan** [comment]
> would like to mount an eval service in an existing server, but don't see a way I can get this working without EventingConfiguration being exported.
> 
> It is possible I am breaking the warranty seal here, but this is pseudo code of what I'm trying to do:
> 
> ```go
> func (f *flagdWrapper) Init(ctx context.Context) error {
> 	log := logging.GetLogger().With("component", "flagd.evaluation.v1").Desugar()
> 	reqIDLogging := false
> 	logger := logger.NewLogger(log, reqIDLogging)
> 	marshalOpts := flagdService.WithJSON(
> 		protojson.MarshalOptions{EmitUnpopulated: true},
> 		protojson.UnmarshalOptions{DiscardUnknown: true},
> 	)
> 	var (
> 		eval            evaluator.IEvaluator
> 		metricsRecorder *telemetry.MetricsRecorder
> 	)
> 	evalSvc := flagdService.NewFlagEvaluationService(
> 		logger, eval, nil /*eventingCfg*/, metricsRecorder,
> 	)
> 	_, flagdHandler := evaluationV1.NewServiceHandler(evalSvc, marshalOpts)
> 
> 	prefix := "/flagd.evaluation.v1.Service/"
> 	service.MustGetHTTPServeMux(ctx).
> 		Handle(prefix, flagdHandler)
> 
> 	return f.Initializer.Init(ctx)
> }
> ```

Thank you for showing the intended usage of the service. I proposed a few improvements to the PR so that it improves testability and makes component interactions clearer.

However, you should understand that as maintainers we cannot guarantee backward compatibility of the `ConnectService` in the long run. On the other hand, we will continue to avoid breaking changes in the flagd core [1] as this is intended to be a reusable component. For example, flagd Go provider use flagd core internally for in-process evaluations. 

[1] - https://github.com/open-feature/flagd/tree/main/core

**Comment by Kavindu-Dodan** [comment]
@connyay let me know if you can work on the proposed changes. If so we can include this alogn with our next release - https://github.com/open-feature/flagd/pull/1292

**Comment by connyay** [comment]
> @connyay let me know if you can work on the proposed changes. If so we can include this alogn with our next release - #1292

I can rework this to use interfaces. Will get to it over the next few days. You dont need to wait on me to cut a release though.

**Comment by Kavindu-Dodan** [comment]
> > @connyay let me know if you can work on the proposed changes. If so we can include this alogn with our next release - #1292
> 
> I can rework this to use interfaces. Will get to it over the next few days. You dont need to wait on me to cut a release though.

Thanks and look forward to the change.

File: flagd/pkg/service/flag-evaluation/connect_service.go
Comment (connyay) [reply flagd/pkg/service/flag-evaluation/connect_service.go]: This introduced some churn without much immediate need to the flagd codebase. I'd like to skip this for now, unless you feel strongly about it.

**Review by Kavindu-Dodan — APPROVED** [review]
Changes looks good 🎉 

@connyay could you please rebase and add DCO signature to commits. You can check the guide here - https://github.com/open-feature/flagd/pull/1288/checks?check_run_id=24131045370

**Comment by connyay** [comment]
> However, you should understand that as maintainers we cannot guarantee backward compatibility of the ConnectService in the long run. On the other hand, we will continue to avoid breaking changes in the flagd core [1] as this is intended to be a reusable component. For example, flagd Go provider use flagd core internally for in-process evaluations.

Totally understand. Appreciate you still being open to this PR even with that being said 👍

**Comment by Kavindu-Dodan** [comment]
> > However, you should understand that as maintainers we cannot guarantee backward compatibility of the ConnectService in the long run. On the other hand, we will continue to avoid breaking changes in the flagd core [1] as this is intended to be a reusable component. For example, flagd Go provider use flagd core internally for in-process evaluations.
> 
> Totally understand. Appreciate you still being open to this PR even with that being said 👍

👍  you're welcome. Please continuously check flagd releases for any important changes. 

I will merge this when other maintainers approve

**Review by beeme1mr — APPROVED** [review]

**Review by toddbaert — APPROVED** [review]
I'm fine with this, given the caveats @Kavindu-Dodan mentioned.

**Pull request merged**
